keps/sig-node/4781-kublet-restart-pod-status/README.md

```diff
@@ -0,0 +1,559 @@
+When the kubelet restarts, it resets the Start and Ready states of all containers to False by default. This means that any successful probe statuses that were previously established are lost upon the restart. As a result, services may be inaccurately flagged as unavailable, despite having been operational prior to the kubelet's restart. This reset can lead to erroneous perceptions of service health and negatively impact the overall performance of the cluster, potentially triggering unnecessary alerts or load balancing changes. 
+It's essential to implement strategies to ensure that the service states accurately reflect their operational status, even during kubelet interruptions.
+
+### Goals
+
+- Ensure consistency in container start and ready states across kubelet restarts.
+- Minimize unnecessary service disruptions caused by temporary ready state changes.
+
+### Non-Goals
+
+- Change the pod not ready status caused by the kubelet restart for over 40 seconds.
+- Modify the fundamental logic of how readiness probes work.
+
+## Proposal
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+As a user of Kubernetes, I want the container's Ready state to remain consistent across kubelet restarts so that my services do not experience unnecessary downtime.
+However, currently, a kubelet restart causes a brief "Not Ready" storm, where the state of all Pods is set to Not Ready, impacting the availability of my services.
+
+### Risks and Mitigations
+
+##### Inconsistency with other Kubernetes components
+If other parts of Kubernetes (e.g., the API server, controllers) expect certain behavior regarding container readiness states, these changes might cause inconsistencies.
+
+##### Delayed Health Check Updates
+By preserving the old state without immediate health checks, there is a delay in recognizing containers that have become unhealthy during or after kubelet's downtime. Services relying on Pod readiness for service discovery might continue directing traffic to Pods with containers that are no longer healthy but are still reported as Ready.
+We plan to immediately trigger a probe after that to reduce the risk caused by such delays.
+
+## Design Details
+
+##### Changes in kubelet_pods.go
+
+We will be updating the `convertToAPIContainerStatuses` function, which is responsible for calculating the API-recognizable v1.ContainerStatus based on the old and current container states. Our modification will involve preserving the Started status from the oldStatus when the container's start time is earlier than the kubelet's start time. This will facilitate the subsequent `UpdatePodStatus` to inherit the Started state.
```

Comment:
I am a little confused about the use of time as an indicator here, but I admit this is not an area of the code I am familiar with.

The way I think about this:

When kubelet starts up, it reads the list of intended pods (and their last known state) from apiserver and it reads the list of actual pods from the CRI.

If it finds a pod in the API that is not running, it starts the pod (yes, I know there's nuance here, especially around state) and updates status.

If it finds a running pod that the API does not have (yes, include mirror pods in that) it kills the pod and cleans up.

If it finds a running pod which corresponds to a pod in the API, it should try to resume the state-machine where the last state indicates.  If it was Ready, it stays ready until the probes fail.  If it was Unready, it stays unready until the probes succeed.  This is the main issue at hand in this KEP, right?

Where does comparing timestamps come in?

Reply:
Your understanding is correct. The comparison of timestamps occurs before checking all containers. It mainly serves two purposes: 
1. to narrow down the comparison scope
2. to avoid cases where the previous container has been restarted. If the container's start time is later than the Kubelet, we will no longer retain the previous state.

Reply:
This is still unclear to me, and I want to make sure the spec is as clear as possible.

> 1. to narrow down the comparison scope

What does this mean?

> 2. to avoid cases where the previous container has been restarted. If the container's start time is later than the Kubelet, we will no longer retain the previous state.

Let me see if I can construct the case you're worried about, please tell me if I am missing it.

t0: container is running and Ready
t1: kubelet goes down
t2: container crashes
t3: CRI impl restarts the container (which is, practically, NotReady)
t4: kubelet comes back up and finds the container running
t5: kubelet does not yet have probe results, so it leaves the container as Ready
t6: eventually the probe shows the container as NotReady

Is that right? If so, I am really not very worried about it.

Referencing myself (sorry) in https://github.com/kubernetes/kubernetes/issues/100277#issuecomment-1179412974 : In a case like this the pod stayed in service a little longer than it should have. This is not a huge deal and happens any time there is a "surprise" change (e.g. the node crashed).

Additionally, the pod start time isn't the only relevant indicator.  That timeline could just as easily be:

t0: container is running and Ready
t1: kubelet goes down
t2: container becomes unready (probes would fail, but app does not crash)
t3: kubelet comes back up and finds the container running
t4: kubelet does not yet have probe results, so it leaves the container as Ready
t5: eventually the probe shows the container as NotReady

There's no timestamp you can compare (I think?) that detects that situation, and the whole reason I think we need to change this behavior is because "assuming the worst" (setting it to NotReady) is way more expensive and problematic than "wait and see".

Upon a kubelet restart, we should probably prioritize "run probes ASAP".

TL;DR - I think that kubelet should only set the status when it actually HAS information, not as a side-effect of anything else.  If you don't KNOW a pod is unready (by probes or some other explicit mechanism) then you shouldn't set it unready.

Reply:
Sorry for not replying recently :(  When updating the pod status, we iterate through all the containers within the pod to identify the containers whose states need to be inherited. Therefore, directly skipping containers with a later start time can narrow down the comparison scope. 
Regarding the first scenario you described and your considerations, I think your understanding is reasonable. It's indeed not a big issue if the pod's state is retained slightly longer. 
In the second scenario, we plan to immediately trigger the probes to let them determine the final state of the pod.

In summary, I think the comparison of the start times can be kept for the purpose of narrowing down the scope. Perhaps I shouldn't have emphasized this point in the KEP. What do you think?